streams: Add stream email address when stream is created.

Kumar-Tarun opened the pull request:
This PR Fixes #8817. We add email_address field to our event, so that on stream/create event, email address is rendered. (earlier it was not rendering).

showell commented on zerver/lib/actions.py:
```diff
@@ -1541,6 +1541,8 @@ def prep_stream_welcome_message(stream: Stream) -> Optional[Dict[str, Any]]:
     event = dict(type="stream", op="create",
                  streams=[stream.to_dict()])
+    # We add email_address to event, since Stream object passed here does not have one.
+    event['streams'][0]['email_address'] = encode_email_address(stream)
```
I think it makes sense to modify a structure **before** putting it in something deeper, so I'd do something like this:

~~~ py
stream_dict = stream.to_dict()
stream_dict['email_address'] = encode_email_address(...)
event = dict(..., streams=[stream_dict])
~~~

This avoids the danger of accidentally associating the wrong email with the wrong stream.

We'll also want a backend test for this.

Kumar-Tarun replied:
Good idea, Thanks. and yes I'm currently taking a look at tests only!

Kumar-Tarun commented:
@showell I've updated the PR, I found another solution to this problem, since subscription/add event was already sending email address, why not use it, so I changed the flow a little bit, and tested it with different scenarios and it works perfectly. You may check! :)

timabbott commented:
I think that solution isn't going to work correctly for some corner cases for organization administrators that @YJDave fixed over the last few months :).   

So I think we just want to the `subscription` event handle to add the email and rerender, not make changes to when `add_sub_to_table` is called.

Kumar-Tarun commented on static/js/server_events_dispatch.js:
```diff
@@ -202,7 +202,6 @@ exports.dispatch_normal_event = function dispatch_normal_event(event) {
             _.each(event.streams, function (stream) {
                 var sub = stream_data.get_sub_by_id(stream.stream_id);
                 stream_data.update_calculated_fields(sub);
-                subs.add_sub_to_table(sub);
```
@timabbott, I am working on this issue, but can you confirm some points below, I reverted my changes, and observed the behavior (normal) of streams for admin:
1) Admin is not able to see email address of stream (both private and public) if he is not subscribed. (Almost same as a normal user with the added ability of seeing private streams (without subscribing), which normal user is not able to see if not invited/subscribed.
2) Another thing, we also do not render email address immediately when stream already exists, and we subscribe afterwards, on refresh it appears.
Thanks for reviewing!

timabbott replied:
Correct.

Kumar-Tarun replied:
If that is the case, then my changes are not breaking the admin experience, right? (because you told those only) :)

timabbott replied:
Yes, but we do need the stream to be added to the "all streams" table for administrators when the stream is created.

Kumar-Tarun replied:
Let me check once again!

Kumar-Tarun replied:
I checked and it (private stream) is being added, email address is only shown when one subscribes (be it admin/general user), the stream is visible in administrator's account in streams table.
This is when I have my changes in effect. :)

YJDave commented on static/js/server_events_dispatch.js:
```diff
@@ -202,7 +202,6 @@ exports.dispatch_normal_event = function dispatch_normal_event(event) {
             _.each(event.streams, function (stream) {
                 var sub = stream_data.get_sub_by_id(stream.stream_id);
                 stream_data.update_calculated_fields(sub);
-                subs.add_sub_to_table(sub);
```
Can you elaborate more about why we want to remove this call from here?

YJDave replied:
This should not be deleted from here. This adds the stream regardless of subscribed or not, to settings when stream is created. If we remove this from here, user will not have stream listed in settings until user is not subscribed to stream or do hard reload.

Kumar-Tarun commented:
This call was creating the stream page without the email address, since it
didn't have the email address field as a part of sub object. I made the
call such that, it is called in the subscription/add event (which anyway
occurs after stream/create event), whose sub object has the email address
field, so it renders correctly on the stream page.
Hope that makes some sense. If you find any case that is being violated, do
let me know (because I have tested and not able to find any), I will be
happy to rectify those! :)



‌

On Thu, Apr 5, 2018 at 9:37 PM, Yashashvi Dave <notifications@github.com>
wrote:

> *@YJDave* commented on this pull request.
> ------------------------------
>
> In static/js/server_events_dispatch.js
> <https://github.com/zulip/zulip/pull/8847#discussion_r179516803>:
>
> > @@ -202,7 +202,6 @@ exports.dispatch_normal_event = function dispatch_normal_event(event) {
>              _.each(event.streams, function (stream) {
>                  var sub = stream_data.get_sub_by_id(stream.stream_id);
>                  stream_data.update_calculated_fields(sub);
> -                subs.add_sub_to_table(sub);
>
> Can you elaborate more about why we want to remove this call from here?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/zulip/zulip/pull/8847#pullrequestreview-109776218>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AWR3I6BAs-aMvMil3YUuMBfJEVNZ_d5rks5tlkFAgaJpZM4S-wex>
> .
>

YJDave commented:
@Kumar-Tarun This solution is not correct, it breaks lots of expected behavior i.e. user should be directed to stream page just after it's creation, it also throws blueslip error. I submitted PR #8991 for fixing this issue, please review PR, so you can get the idea of expected behavior and can make suggestions.

YJDave commented on static/js/server_events_dispatch.js:
```diff
@@ -235,6 +234,7 @@ exports.dispatch_normal_event = function dispatch_normal_event(event) {
         if (event.op === 'add') {
             _.each(event.subscriptions, function (rec) {
                 var sub = stream_data.get_sub_by_id(rec.stream_id);
+                sub.email_address = rec.email_address;
```
We don't update `sub` from this file, `stream_data.js` is right place to add this code.

Kumar-Tarun replied:
Agreed!

YJDave commented on static/js/subs.js:
```diff
@@ -222,15 +222,14 @@ exports.update_settings_for_subscribed = function (sub) {
         settings_button.text(i18n.t("Unsubscribe"));
 
         stream_edit.add_me_to_member_list(sub);
-    } else {
```
This else shouldn't be removed, this checks if subscription template is already rendered or not. Same way `add_sub_to_table` is also shouldn't called every time `update_settings_for_subscribed` gets called.

timabbott commented:
Closing in favor of #8991.
